### PR TITLE
Ip/nac 1657

### PR DESCRIPTION
**Pull request by ipiatrou**
## What’s changing?

BulkUpload azure function has been updated.
Updated the logic to validate CVS file.
Now a user can see all errors per row in the uploaded file.
All errors from the uploaded file you can see in piipan.uploads table "error_messages" or in azure logs.

To test my changes I have set a value 30000 for **max_prepared_transactions** for psql-participants-dev server + restart the server.
It is a temporary solution. Bill is working to fix it. 
MicrosoftTeams-image (4)

The error message looks like this:
image

## Why?

NAC-1657

## This PR has:

- [x] **JIRA ticket(s)** (_Must be listed in PR request and branch name should follow convention NAC-123-description_)<description>

- [ ] **Security Related Changes** (_Security Tab questions have been answered on JIRA ticket(s)_)

- [x] **Commented source code** (_required on public classes and methods, and elsewhere as appropriate_)

- [ ] **Documented API changes** (_required on public classes and methods, and elsewhere as appropriate_)
    - [ ] External API documentation - OpenApi docs and Generated Docs (e.g. https://github.com/18F/piipan/tree/dev/docs/openapi) 
    - [ ] Internal API documentation (e.g. https://github.com/18F/piipan/tree/dev/metrics/docs/openapi)   

- [ ] **Developer documentation** (_for new build/test/API changes or complex portions of the system_)

- [x] **Automated unit tests** (_to maintain or increase level of code coverage_)

- [ ] **Changes to IAC** (_add any deployment steps that require manual intervention to the draft release notes_)
    - [ ] Have IaC scripts run successfully against an existing deployed environment?
    - [ ] Have IaC scripts run successfully against a newly deployed environment?
    - [ ] Do the IaC changes require deleting existing resources?

## Anything else?

_Add other details that are helpful for review here_

## Code Review Checklist:

- [ ] Automated checks all pass? (code builds successfully, tests pass, test coverage meets/exceeds threshold, etc.) 
- [ ] Have all requirements and Acceptance Criteria been met?
- **If there are code changes**
    - [ ] Is there no unnecessary complexity? Code is readable, maintainable, etc.
    - [ ] Do the changes follow SOLID principles?
    - [ ] Are the changes robust? Concurrency considered? Performance? Security (e.g. no sql injections)?
    - [ ] Are newly added dependencies needed and appropriate? Are dependency files (e.g. packages.json & packages.lock.json) updated?
    - [ ] For any new projects/solutions, have they been added to code coverage files?
    - [ ] Public classes & methods are documented    
- **If there are API changes**
    - [ ] Does API make sense? Useful and concise?
    - [ ] Are existing API interfaces/contracts intact? i.e. no breaking changes
    - [ ] Are API tests included?
    - [ ] Are API responses appropriate and documented? (200s, 201, 404s, etc.)
    - [ ] Are internals kept private and not leaked into the API?
    - [ ] External API documentation - OpenApi docs and Generated Docs 
    - [ ] Internal API documentation    
- **If there are IAC changes**
    - [ ] IAC Scripts are readable & maintainable
    - [ ] IAC Scripts are documented
    - [ ] IAC scripts pass required checks
    - [ ] IAC scripts were run successfully by author
- **General Logging Questions**
    - [ ] Do new capabilities (or changed capabilities) include logging?
    - [ ] Is PII not logged?
- **General Error Handling Questions**
    - [ ] Has exception handling been added for new capabilities (or changed capabilities)?
    - [ ] Are errors/exceptions properly communicated to end-users? Were there requirements for user facing error messages?
    - [ ] Are errors/exceptions logged?
    - [ ] Is PII kept out of error/exception logging?
- **If there are Developer Documentation Changes**
    - [ ] General Project documentation (under piipan\docs. e.g. iac.md, architecture.md, etc.)
    - [ ] Github documentation additions/changes (Readmes)
- **General automated test questions**
    - [ ] End-to-End Tests provided (if code implements new user story)
    - [ ] Are tests sufficient?
    - [ ] Corner cases considered?
    - [ ] Are performance tests needed?

### CODE REVIEW

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
-using System;
-using System.IO;
-using System.Threading.Tasks;
 using Microsoft.Azure.WebJobs;
 using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
 using Piipan.Etl.Func.BulkUpload.Parsers;
+using Piipan.Etl.Func.BulkUpload.Validators;
 using Piipan.Participants.Api;
+using System;
```
Comment (bgfoshay) [inline etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs:9]: Nitpick: Can you configure VS to organize your using statements with System ones listed first?

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs [hunk 2]
```diff
@@ -65,28 +70,44 @@ public class BulkUpload
...
-                        var participants = _participantParser.Parse(input);
-                        DateTime startUploadTime = DateTime.UtcNow;
-
                         //Azure documents that ETags are quoted. So we need to remove the quotes in order to get the upload_id
                         var upload_id = blobProperties.ETag.ToString().Replace("\"", "");
-
                         string state = Environment.GetEnvironmentVariable("State");
+                        DateTime startUploadTime = DateTime.UtcNow;
+
+                        var validationReport = _csvValidator.ValidateCsvDoc(input, blockBlobClient.Name, upload_id);
```
Comment (bgfoshay) [inline etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs:78]: Question: Does this mean we're now doing two passes through the csv file?

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Parsers/IParticipantStreamParser.cs [hunk 3]
```diff
@@ -7,5 +7,6 @@ namespace Piipan.Etl.Func.BulkUpload.Parsers
     public interface IParticipantStreamParser
     {
         IEnumerable<IParticipant> Parse(Stream input);
+        HashSet<string> GetPLLInformation(Stream input);
```
Comment (dborgen2) [inline etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Parsers/IParticipantStreamParser.cs:10]: I think PLL is supposed to be PII (Personally Identifiable Information) here and in the other places

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/Helper.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.Globalization;
+
+namespace Piipan.Etl.Func.BulkUpload.Validators
+{
+    public class Helper
```
Comment (dborgen2) [inline etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/Helper.cs:6]: Can the name of this class be a bit more descriptive, such as ValidatorHelper? Also, since it's a new class we should have comments on it and all public methods specifying their uses. Not all classes/methods have that right now, but we've been trying to get better at it and we get flagged for it if we don't

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs (hunk 1 above)
Comment (ipiatrou) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs]: updated

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Models/ParticipantCsv.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+﻿using CsvHelper.Configuration.Attributes;
+
+namespace Piipan.Etl.Func.BulkUpload.Models
+{
+    public class ParticipantCsv
```
Comment (bgfoshay) [inline etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Models/ParticipantCsv.cs:5]: Nitpick: We need to comment blocks documenting our public classes and public methods

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs (hunk 2 above)
Comment (ipiatrou) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs:78]: yep.
The first time for error validation for each row + creating a validation report.
The second time to store it in the database.

Comment (bgfoshay) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs:78]: Ok. I know you and I discussed a little bit. But you're not seeing a significant performance impact in your testing for this, right? (when testing with ~1-2 million participants).

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Parsers/ParticipantStreamCsvParser.cs [hunk 6]
```diff
@@ -207,5 +161,39 @@ public IEnumerable<IParticipant> Parse(Stream input)
             // Yields records as it is iterated over
             return csv.GetRecords<Participant>();
         }
+
+        public HashSet<string> GetPLLInformation(Stream input)
+        {
+            var hashSet = new HashSet<string>();
+
+            var reader = new StreamReader(input);
```
[the remark this answers is not in the document]
Comment (bgfoshay) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Parsers/ParticipantStreamCsvParser.cs:169]: Question: Is this doing a third pass through the csv? One pass for validation, one for loading the hashset, and one for reading records and persisting to the DB?

File: etl/tests/Piipan.Etl.Func.BulkUpload.Tests/CsvValidatorTest.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+            //Assert
+            Assert.NotNull(res);
+            Assert.True(res.IsValid);
+            Assert.True(res.Errors.Count == 0);
+            Assert.True(res.FileName == fileName);
+            Assert.True(res.ETag == eTag);
+        }
+        [Fact]
+        public void ValidateCsvDoc_SouldReturnErrors()
```
Comment (bgfoshay) [inline etl/tests/Piipan.Etl.Func.BulkUpload.Tests/CsvValidatorTest.cs:29]: Nitpick: typo in test name

File: participants/src/Piipan.Participants/Piipan.Participants.Core/DataAccessObjects/UploadDao.cs [hunk 8]
```diff
@@ -54,7 +54,7 @@ public async Task<IUpload> AddUpload(string uploadIdentifier)
         {
             try
             {
-                using (var connection = await _dbConnectionFactory.Build() as DbConnection)
+                using (var connection = await _dbConnectionFactory.BuildConnectionWithNodaTime() as DbConnection)
```
Comment (bgfoshay) [inline participants/src/Piipan.Participants/Piipan.Participants.Core/DataAccessObjects/UploadDao.cs:57]: Thought: Just flagging that this will likely conflict with Dan's database changes to address Datasource and NodaTime bugs

File: participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs [hunk 9]
```diff
@@ -121,53 +118,63 @@ public async Task AddParticipants(IEnumerable<IParticipant> participants, string
...
+                var redactedErrorMessage = redacteErrorMessageCallback?.Invoke(ex);
+                await LogParticipantsUploadError(upload, participantUploadMetrics, redactedErrorMessage);
+            }
+        }
 
-                upload.Status = UploadStatuses.FAILED.ToString();
-                upload.ErrorMessage = ex.Message;
-                upload.CompletedAt = failureTime;
-                await _uploadDao.UpdateUpload(upload);
+        public async Task LogParticipantsUploadError(string errorMessage, string uploadIdentifier, string state)
```
Comment (bgfoshay) [inline participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs:126]: Question: Is this method being used? Also, is the responsibility of this method for "logging" or is it "recording errors in db"?

File: participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs [hunk 10]
```diff
@@ -121,53 +118,63 @@ public async Task AddParticipants(IEnumerable<IParticipant> participants, string
...
+            participantUploadMetrics.CompletedAt = failureTime;
+            participantUploadMetrics.ErrorMessage = exceptionMessage;
+
+            string uploadDBStatus = JsonConvert.SerializeObject(participantUploadMetrics);
+
+            await _participantPublishUploadMetric.PublishUploadMetric(participantUploadMetrics);
         }
 
-        public void LogParticipantsUploadError(ParticipantUploadErrorDetails errorDetails, IEnumerable<IParticipant> participants)
+        public string RedacteParticipantsUploadError(ParticipantUploadErrorDetails errorDetails, HashSet<string> pllInfo)
```
Comment (bgfoshay) [inline participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs:159]: Question: Could this method be moved out of this class to RedactionService.cs?

Comment (bgfoshay) [inline participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs:159]: Nitpick: Typo in method name-> "Redact"

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs [hunk 11]
```diff
@@ -65,28 +70,44 @@ public class BulkUpload
...
+                            return;
+                        }
+
+                        var participants = _participantParser.Parse(input);
+
+                        await _participantApi.AddParticipants(participants, upload_id, state, (ex) => LogAndRedacteErrors(ex))
                                 .ContinueWith(t => _blobStream.DeleteBlobAfterProcessing(t, blockBlobClient, log))
                                 .ContinueWith(t => _participantApi.DeleteOldParticpants());
 
+                        string LogAndRedacteErrors(Exception ex)
```
Comment (bgfoshay) [inline etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs:98]: Nitpick: minor typo in method name -> "Redact"

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs [hunk 12]
```diff
@@ -65,28 +70,44 @@ public class BulkUpload
...
                         var upload_id = blobProperties.ETag.ToString().Replace("\"", "");
-
                         string state = Environment.GetEnvironmentVariable("State");
+                        DateTime startUploadTime = DateTime.UtcNow;
+
+                        var validationReport = _csvValidator.ValidateCsvDoc(input, blockBlobClient.Name, upload_id);
+
+                        if (!validationReport.IsValid)
+                        {
+                            var errorMessage = JsonConvert.SerializeObject(validationReport);
```
Comment (bgfoshay) [inline etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs:82]: Thought: Given this error message is really serialized version of the validation report, would it make the code any clearer to refer to it as something like "serializedValidationErrors" or "serializedValidationReport", both for these variables and in the LogParticipantsUploadError method parameter?

File: participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs (hunk 9 above)
Comment (bgfoshay) [reply participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs:126]: Also, given it's creating a "failure" upload record and recording the validation errors to that record, maybe it should be something like "CreateUploadFailureRecordWithValidationErrors"?

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs (hunk 2 above)
Comment (ipiatrou) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs:78]: I uploaded not valid CSV with 1 million records.
It took 58 sec to validate it and sent back the validation report.

image

It took 3.7 min to validate and store in the database 1 million valid rows.

image

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Parsers/ParticipantStreamCsvParser.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ipiatrou) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Parsers/ParticipantStreamCsvParser.cs:169]: Yes. It must have using block.

we have 3 cases:

1. cvs file is not valid - 1 pass just to get validation errors
2. cvs file is valid - 1 pass just to make sure it is valid + 1 pass to store it in the database
3. cvs file is valid but throws an exception - 1 pass just to make sure it is valid + 1 pass to store it in the database (it can be aborted) + 1 pass to make sure the exception message does not contain PII **(very rare case)**

File: participants/src/Piipan.Participants/Piipan.Participants.Core/DataAccessObjects/UploadDao.cs (hunk 8 above)
Comment (ipiatrou) [reply participants/src/Piipan.Participants/Piipan.Participants.Core/DataAccessObjects/UploadDao.cs:57]: ok.
without this fix, I could not test my changes on dev env.

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/CsvValidator.cs [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        public CsvValidator(IValidator<ParticipantCsv> validator)
+        {
+            _validator = validator;
+        }
+
+        public CsvValidationReport ValidateCsvDoc(Stream input, string fileName, string eTag)
+        {
+            var report = new CsvValidationReport(fileName, eTag, true);
+
+            var reader = new StreamReader(input);
```
[the remark this answers is not in the document]
Comment (ipiatrou) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/CsvValidator.cs:29]: We have "using" statement into  BulkUpload class, method Run, line 70

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs (hunk 12 above)
Comment (ipiatrou) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs]: Yep. I have updated it

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs (hunk 11 above)
Comment (ipiatrou) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/BulkUpload.cs]: Fixed

File: participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs (hunk 10 above)
Comment (ipiatrou) [reply participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs]: Fixed

Comment (ipiatrou) [reply participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs]: Done. Make seans

File: participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs (hunk 9 above)
Comment (ipiatrou) [reply participants/src/Piipan.Participants/Piipan.Participants.Core/Services/ParticipantService.cs]: Fixed

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/CsvValidator.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (ipiatrou) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/CsvValidator.cs:30]: @dborgen2 
That is an interesting suggestion, I did not use it before. 
I played around with leaveOpen in ParticipantCsvStreamParser.Parse method but it does not work as expected. Also, I set up leaveOpen: true into CsvReader constructor but it does not work as well.

I did some test:
image

The result: 
17 tests failed (Tested locally).
Error message: 
**System.ObjectDisposedException: 'Cannot read from a closed TextReader. Object name: 'StreamReader'.'**

I do not see a way how to add "using" statement to StreamReader because we are using Read method from  StreamReader inside CsvReader.
I think we are good because we have "using" statement in BulkUpload class

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Models/ParticipantCsv.cs (hunk 5 above)
Comment (bgfoshay) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Models/ParticipantCsv.cs:5]: We need to document public classes and methods like the following example 

/// <summary>
/// Class <c>Point</c> models a point in a two-dimensional plane.
/// </summary>
public class Point
{
    /// <summary>
    /// Method <c>Draw</c> renders the point.
    /// </summary>
    void Draw() {...}
}

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Models/CsvValidationReport.cs [hunk 14]
```diff
@@ -0,0 +1,29 @@
+﻿using System.Collections.Generic;
+using Newtonsoft.Json;
+
+namespace Piipan.Etl.Func.BulkUpload.Models
+{
+    public class CsvValidationReport
```
Comment (bgfoshay) [inline etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Models/CsvValidationReport.cs:6]: Nitpick: We need documentation for the class and methods here too.

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/CsvValidator.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (bgfoshay) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/CsvValidator.cs:30]: Let's chat about this on Monday. I agree it should be covered by the using statement in the BulkUpload class. But I'm a little nervous about an open stream being passed around. If there's a way we can come up with to encapsulate the stream usage, it would lower the risk of this class being misused by someone else in the future.

File: etl/tests/Piipan.Etl.Func.BulkUpload.Tests/CsvValidatorTest.cs [hunk 15]
```diff
@@ -0,0 +1,130 @@
+﻿using System.IO;
+using System.Linq;
+using Piipan.Etl.Func.BulkUpload.Validators;
+using Xunit;
+
+namespace Piipan.Etl.Func.BulkUpload.Tests
+{
+    public class CsvValidatorTest
```
Comment (bgfoshay) [inline etl/tests/Piipan.Etl.Func.BulkUpload.Tests/CsvValidatorTest.cs:8]: Praise: Thanks for being thorough covering all these different test cases!

File: etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/CsvValidator.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (bgfoshay) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/CsvValidator.cs:30]: Also, I just want to acknowledge that this pattern existed before this PR in the ParticipantStreamCsvParser. It's not new. I'm just wondering if there's a way to tidy up while we're looking at it.

[the remark this answers is not in the document]
Comment (dborgen2) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/CsvValidator.cs:30]: I'm good with getting on a call about this. But yes @ipiatrou, that will fail because the using is scoped to that block, and will be disposed when you leave the block.

I think
`using var reader = new StreamReader(input, leaveOpen: true)` should work, as it would get disposed when the function completes. But again, I agree with Bill, we should probably all get on the same page with what we want open vs what we don't.

[the remark this answers is not in the document]
Comment (ipiatrou) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/CsvValidator.cs:30]: @dborgen2 
Unfortunately, it will not work because we are using IEnumerable<IParticipant> response and we do not call Read method inside CSVHelper into the Parse method. 
In the different method, before storing in the database, we are getting all participants. At this moment we call method (Read) inside of CSVHelper but this method is not available due to "using" statement.

This principle is represented in the BadCSVError unit test for example (and +16 unite tests). If we are going to use "using" statement 17 tests will be filed. 
The main issue is passing stream in a couple of methods.

[the remark this answers is not in the document]
Comment (dborgen2) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/CsvValidator.cs:30]: Gotcha, thanks for the explanation. I feel like there's still got to be a way to do it, but it would probably involve disposing the CsvValidator which would in turn dispose the StreamReader (assuming the StreamReader was a class variable). Not sure if it's worth doing all of that or not. According to the [StreamReader Dispose doc](https://learn.microsoft.com/en-us/dotnet/api/system.io.streamreader.dispose?view=net-7.0), in additional to closing the stream (if leaveOpen is false), it "releases the unmanaged resources used by the [StreamReader](https://learn.microsoft.com/en-us/dotnet/api/system.io.streamreader?view=net-7.0), and optionally releases the managed resources". So that's kind of vague, but I know in general it's bad practice to not dispose of things like this.

Maybe when we get on a call about passing the stream around we should brainstorm this problem.

[the remark this answers is not in the document]
Comment (dborgen2) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/CsvValidator.cs:30]: Just to follow-up, we decided since this is a class based on existing functionality (ParticipantStreamCsvParser), we should be consistent with that, and that we will have another story for researching whether or not both of these classes should be refactored. So this is good for now.

[the remark this answers is not in the document]
Comment (ipiatrou) [reply etl/src/Piipan.Etl/Piipan.Etl.Func.BulkUpload/Validators/CsvValidator.cs:30]: Sounds good

**Review by bgfoshay — APPROVED** [review]
👍 Thanks for all the work that you did with this one! Thanks for entertaining my questions around the stream/reader usage. I also know there was a lot of nitpicks around documentation and I appreciate you taking the time to add it. States are definitely going to benefit from the additional error detail this work provides.
